model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanSamlArtifactSessionMappingStoreProvider.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.models.sessions.infinispan;
+
+import org.infinispan.client.hotrod.exceptions.HotRodClientException;
+import org.infinispan.commons.api.BasicCache;
+import org.jboss.logging.Logger;
+import org.keycloak.common.util.Time;
+import org.keycloak.models.AuthenticatedClientSessionModel;
+import org.keycloak.models.SamlArtifactSessionMappingModel;
+import org.keycloak.models.SamlArtifactSessionMappingStoreProvider;
+import org.keycloak.models.sessions.infinispan.entities.ActionTokenValueEntity;
+import org.keycloak.models.sessions.infinispan.util.InfinispanUtil;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+
+/**
+ * @author mhajas
+ */
+public class InfinispanSamlArtifactSessionMappingStoreProvider implements SamlArtifactSessionMappingStoreProvider {
+
+    public static final Logger logger = Logger.getLogger(InfinispanSamlArtifactSessionMappingStoreProvider.class);
+
+    private static final String USER_SESSION_ID = "user.session.id";
+    private static final String CLIENT_SESSION_ID = "client.session.id";
+
+    private final Supplier<BasicCache<UUID, ActionTokenValueEntity>> codeCache;
```

Comment:
why is this cache here indirectly via `Supplier`?
Could it be declared as 
```suggestion
    private final BasicCache<String, String[]> codeCache;
```
where key would be artifact ID and value would be `new String[] {USER_SESSION_ID, CLIENT_SESSION_ID}`?

Reply:
This is a good question. I just used the same thing as was used for CodeToToken cache. @mposolda is the Supplier necessary there? See the same part for CodeToToken provider: https://github.com/keycloak/keycloak/blob/master/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanCodeToTokenStoreProvider.java#L41 

I tried to call `get()` method when  `InfinispanSamlArtifactSessionMappingStoreProvider` is created and it works, but I didn't test it in clustered/crossDc environment.

Reply:
AFAIK The reason is this https://github.com/keycloak/keycloak/blob/master/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanSingleUseTokenStoreProviderFactory.java#L71 as this is the per-invocation flag in remoteCache to FORCE_RETURN_VALUE. At least at the time when we worked on this it was per-invocation flag (but it is a chance that this is changed in latest infinispan version). This might be needed for the operations, which are supposed to return something. Like for example `putIfAbsent` or `remove`.

Reply:
I see, thanks for the explanation! That makes sense, let's keep it as is for now.